presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrder.java

```diff
@@ -284,36 +283,88 @@ private void checkEncodeInputValidity(List<Integer> input, int maximumBits)
         checkArgument((totalBitLength <= maximumBits), format("The z-address type specified is not large enough to hold %d bits.", totalBitLength));
     }
 
+    /**
+     * The BoundedZValueRange class contains non-null Integer minimum and maximum value lists. The bounds are all inclusive.
+     */
+    private static class BoundedZValueRange
+    {
+        private final List<Integer> minimumValues;
+        private final List<Integer> maximumValues;
+
+        public BoundedZValueRange(List<Integer> minimumValues, List<Integer> maximumValues)
+        {
+            this.minimumValues = minimumValues;
+            this.maximumValues = maximumValues;
+        }
+
+        public Integer getMinimumValueAt(int index)
+        {
+            return minimumValues.get(index);
+        }
+
+        public Integer getMaximumValueAt(int index)
+        {
+            return maximumValues.get(index);
+        }
+    }
+
     private List<ZAddressRange<Long>> zOrderSearchCurve(List<ZValueRange> valueRanges)
     {
-        List<ZAddressRange<Long>> addressRanges = new ArrayList<>();
+        List<BoundedZValueRange> boundedValueRanges = fillUnspecifiedRangesWithDefaults(valueRanges);
 
-        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, fillUnspecifiedRangesWithDefaults(valueRanges), addressRanges);
+        int[] lengths = new int[valueRanges.size()];
+        for (int index = 0; index < valueRanges.size(); index++) {
+            lengths[index] = valueRanges.get(index).length;
+        }
+
+        int[] counters = new int[valueRanges.size()];
+        Arrays.fill(counters, 0);
+
+        List<ZAddressRange<Long>> addressRanges = new ArrayList<>();
+        permuteValueRanges(counters, lengths, 0, boundedValueRanges, addressRanges);
 
         return combineAddressRanges(addressRanges);
     }
 
-    private List<ZValueRange> fillUnspecifiedRangesWithDefaults(List<ZValueRange> valueRanges)
+    private List<BoundedZValueRange> fillUnspecifiedRangesWithDefaults(List<ZValueRange> valueRanges)
     {
-        List<ZValueRange> nonNullValues = new ArrayList<>(valueRanges.size());
+        List<BoundedZValueRange> nonNullValues = new ArrayList<>(valueRanges.size());
         for (int index = 0; index < valueRanges.size(); index++) {
             ZValueRange range = valueRanges.get(index);
 
-            Integer minimumValue = range.getMinimumValue();
-            Integer maximumValue = range.getMaximumValue();
-
-            if (minimumValue == null) {
-                minimumValue = 0;
+            List<Integer> minimumValues = new ArrayList<>();
+            for (Integer minimumValue : range.getMinimumValues()) {
+                if (minimumValue == null) {
+                    minimumValue = 0;
+                }
```

Comment:
not so comfortable handling nulls outside of `getMinimumValues` (or max). Can we make sure the list is with all non-null values? Do we expect special meanings for nulls vs 0?

Reply:
Yes, `null` is used for partial range queries. For example, if the query asks for `column <= 5`, the range will be [null, 5]. In this implementation, after we know `encodingBits`, we can set default values: 0 for minimum, and `(1 << encodingBits) - 1` for maximum.

One thing I can do is pass in `encodingBits` to have the nulls them handled inside the `ZValueRange.getMinimumValues` function instead of outside.